### PR TITLE
drivers/mpl3115a2: rework

### PR DESCRIPTION
**Pull request by smlng**
- rework driver interface
- adapt test
- add saul support -> i.e., for pba-d-01-kw2x

### CODE REVIEW

File: drivers/mpl3115a2/include/mpl3115a2.h [hunk 1]
```diff
@@ -38,6 +40,16 @@ extern "C"
 {
 #endif
 
+/**
+ * @brief   Named return values
+ */
+enum {
+    MPL3115A2_OK,                 /**< all good */
+    MPL3115A2_ERROR_I2C,          /**< I2C communication failed */
+    MPL3115A2_ERROR_DEV,          /**< Device MAG3110 not found */
```
Comment (smlng) [inline drivers/mpl3115a2/include/mpl3115a2.h]: note to self: typo /MAG3110/MPL3115A2/

File: drivers/mpl3115a2/include/mpl3115a2_params.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+#endif
+#ifndef MPL3115A2_PARAMS
+#define MPL3115A2_PARAMS        { .i2c  = MPL3115A2_PARAM_I2C, \
+                                  .addr = MPL3115A2_PARAM_ADDR, \
+                                  .rate = MPL3115A2_OS_RATIO_DEFAULT }
+#endif
+/**@}*/
+
+/**
+ * @brief   MAG3110 configuration
```
Comment (smlng) [inline drivers/mpl3115a2/include/mpl3115a2_params.h]: note to self: typo again

**Comment by smlng** [comment]
see #7152, too

**Comment by smlng** [comment]
anybody willing to review, this is would finally resolve (and close) issue #7152

File: drivers/mpl3115a2/include/mpl3115a2_params.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ * @{
+ */
+#ifndef MPL3115A2_PARAM_I2C
+#define MPL3115A2_PARAM_I2C         (I2C_DEV(0))
+#endif
+#ifndef MPL3115A2_PARAM_ADDR
+#define MPL3115A2_PARAM_ADDR        (MPL3115A2_I2C_ADDRESS)
+#endif
+#ifndef MPL3115A2_PARAMS_DEFAULT
+#define MPL3115A2_PARAMS_DEFAULT    {   .i2c  = MPL3115A2_PARAM_I2C, \
```
Comment (aabadie) [inline drivers/mpl3115a2/include/mpl3115a2_params.h]: extra spaces before `.i2c`, one is enough

File: drivers/mpl3115a2/mpl3115a2.c [hunk 4]
```diff
@@ -16,211 +17,180 @@
...
-    i2c_release(dev->i2c);
-
-    if (mpl3115a2_test(dev)) {
-        return -3;
+    /* test device */
+    if (i2c_read_regs(BUS, ADDR, MPL3115A2_WHO_AM_I, &reg, 1) != 1) {
+        /* Release the bus for other threads. */
+        i2c_release(BUS);
+        LOG_ERROR("mpl3115a2_init: I2C error!\n");
+        return -MPL3115A2_ERROR_I2C;
```
Comment (aabadie) [inline drivers/mpl3115a2/mpl3115a2.c]: Maybe a specific enum for I2C read error could be used ?

File: drivers/mpl3115a2/mpl3115a2.c [hunk 5]
```diff
@@ -16,211 +17,180 @@
...
-    i2c_acquire(dev->i2c);
-    if (i2c_write_regs(dev->i2c, dev->addr, MPL3115A2_PT_DATA_CFG, &reg, 1) != 1) {
-        i2c_release(dev->i2c);
-        return -4;
+    /* set sample rate */
+    reg = MPL3115A2_CTRL_REG1_OS(dev->params.rate);
+    if (i2c_write_regs(BUS, ADDR, MPL3115A2_CTRL_REG1, &reg, 1) != 1) {
+        i2c_release(BUS);
+        LOG_ERROR("mpl3115a2_init: failed to set sample rate!\n");
+        return -MPL3115A2_ERROR_CNF;
```
Comment (aabadie) [inline drivers/mpl3115a2/mpl3115a2.c]: If the I2C read error code is kept, this one could be renamed to MPL3115A2_ERROR_WRITE. What do you think ?

File: drivers/mpl3115a2/mpl3115a2_saul.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (C) 2017 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     driver_mpl3115a2
```
Comment (aabadie) [inline drivers/mpl3115a2/mpl3115a2_saul.c]: driver**s**_mpl3115a2

File: drivers/include/mpl3115a2.h [hunk 7]
```diff
@@ -118,7 +125,7 @@ int mpl3115a2_set_active(mpl3115a2_t *dev);
  *
  * @param[in]  dev          device descriptor of sensor
  *
- * @return                  0 on success
+ * @return                  0 (MPL3115A2_OK) on success
  * @return                  -1 on error
  */
 int mpl3115a2_set_standby(mpl3115a2_t *dev);
```
Comment (aabadie) [inline drivers/include/mpl3115a2.h]: Please make the device descriptors `const` when needed/possible.

**Comment by smlng** [comment]
@aabadie: I addressed your comments.

File: drivers/include/mpl3115a2.h [hunk 8]
```diff
@@ -38,78 +38,85 @@ extern "C"
...
-#define MPL3115A2_OS_RATIO_16           4 /**< Oversample Ratio 16, conversion time 66 ms */
-#define MPL3115A2_OS_RATIO_32           5 /**< Oversample Ratio 32, conversion time 130 ms */
-#define MPL3115A2_OS_RATIO_64           6 /**< Oversample Ratio 64, conversion time 258 ms */
-#define MPL3115A2_OS_RATIO_128          7 /**< Oversample Ratio 128, conversion time 512 ms */
-#define MPL3115A2_OS_RATIO_DEFAULT      MPL3115A2_OS_RATIO_128 /**< Default Ratio for testing */
+/**
+ * @name    Oversample Ratio configuration
+ * @{
+ */
+#define MPL3115A2_OS_RATIO_1        (0x00) /**< Oversample Ratio 1, conversion time 6 ms */
```
Comment (aabadie) [inline drivers/include/mpl3115a2.h]: Maybe this can be in a anonymous enum?

File: drivers/include/mpl3115a2.h [hunk 9]
```diff
@@ -38,78 +38,85 @@ extern "C"
...
+ * @brief   Configuration parameters
  */
 typedef struct {
-    i2c_t i2c;              /**< I2C device, the sensor is connected to */
-    uint8_t addr;           /**< the sensor's slave address on the I2C bus */
-    bool initialized;       /**< sensor status, true if sensor is initialized */
-} mpl3115a2_t;
+    i2c_t i2c;                  /**< I2C bus the device is connected to */
+    uint8_t addr;               /**< I2C bus address of the device */
+    uint8_t rate;               /**< sampling rate */
```
Comment (aabadie) [inline drivers/include/mpl3115a2.h]: is it oversample ratio ? I would call it oversampling ratio btw

**Comment by smlng** [comment]
@aabadie happy? Regarding your comment on i2c read/write errors and matching return value, we don't have that for any other i2c driver. I see you point that this distinction might be useful, however, IMO we should generalize i2c error codes for return anyway - that is: define them in the i2c header file instead of redefinition for each and every driver.

**Comment by aabadie** [comment]
> @aabadie happy?

Yes !

**Comment by aabadie** [comment]
There are lots of things that could be improved/modified with drivers implementation.

**Comment by aabadie** [comment]
and go!

**Comment by smlng** [comment]
> There are lots of things that could be improved/modified with drivers implementation.

yes, I2C is missing a major overhaul - though, @haukepetersen is/was already working on this

**Pull request merged**